File: contest_8/B.py
from sys import setrecursionlimit
from math import inf

setrecursionlimit(100000)


class SegmentTree:
    def __init__(self, elements):
        self.tree = [None] * (4 * n)
        self.build(elements, 0, 0, n - 1)

    def build(self, arr, x, tl, tr):
        if tl == tr:
            self.tree[x] = (arr[tl], 1)
        else:
            tm = (tl + tr) // 2
            self.build(arr, x * 2 + 1, tl, tm)
            self.build(arr, x * 2 + 2, tm + 1, tr)
            self.tree[x] = self.merge(self.tree[x * 2 + 1], self.tree[x * 2 + 2])

    @staticmethod
    def merge(left, right):
        if left[0] == right[0]:
            return left[0], left[1] + right[1]
        elif left[0] < right[0]:
            return left
        else:
            return right

    def min(self, x, tl, tr, left, right):
        if left > right:
            return inf, 0
        if left == tl and right == tr:
            return self.tree[x]
        tm = (tl + tr) // 2
        return self.merge(
            self.min(x * 2 + 1, tl, tm, left, min(right, tm)),
            self.min(x * 2 + 2, tm + 1, tr, max(left, tm + 1), right)
        )

    def update(self, x, tl, tr, ind, value):
        if tl == tr:
            self.tree[x] = (value, 1)
        else:
            tm = (tl + tr) // 2
            if ind <= tm:
                self.update(x * 2 + 1, tl, tm, ind, value)
            else:
                self.update(x * 2 + 2, tm + 1, tr, ind, value)
            self.tree[x] = self.merge(self.tree[x * 2 + 1], self.tree[x * 2 + 2])


n, m = map(int, input().split())
lst = list(map(int, input().split()))
seg_tree = SegmentTree(lst)
for _ in range(m):
    query, *args = map(int, input().split())
    if query == 1:
        i, v = args
        seg_tree.update(0, 0, n - 1, i, v)
    elif query == 2:
        l, r = args
        print(*seg_tree.min(0, 0, n - 1, l, r - 1))

File: contest_8/B.cpp
#include <vector>
#include <climits>
#include <stdio.h>
#pragma GCC optimize("Ofast")

using namespace std;

class SegmentTree {
private:
    vector<pair<int, int>> tree;
    int n;

    pair<int, int> merge(const pair<int, int>& left, const pair<int, int>& right) {
        if (left.first == right.first) {
            return {left.first, left.second + right.second};
        } else if (left.first < right.first) {
            return left;
        } else {
            return right;
        }
    }

    void build(const vector<int>& arr, int x, int tl, int tr) {
        if (tl == tr) {
            tree[x] = {arr[tl], 1};
        } else {
            int tm = (tl + tr) / 2;
            build(arr, x * 2 + 1, tl, tm);
            build(arr, x * 2 + 2, tm + 1, tr);
            tree[x] = merge(tree[x * 2 + 1], tree[x * 2 + 2]);
        }
    }

    pair<int, int> min_cnt(int x, int tl, int tr, int left, int right) {
        if (left > right) {
            return {INT_MAX, 0};
        }
        if (left == tl && right == tr) {
            return tree[x];
        }
        int tm = (tl + tr) / 2;
        return merge(
            min_cnt(x * 2 + 1, tl, tm, left, min(right, tm)),
            min_cnt(x * 2 + 2, tm + 1, tr, max(left, tm + 1), right)
        );
    }

    void update(int x, int tl, int tr, int ind, int value) {
        if (tl == tr) {
            tree[x] = {value, 1};
        } else {
            int tm = (tl + tr) / 2;
            if (ind <= tm) {
                update(x * 2 + 1, tl, tm, ind, value);
            } else {
                update(x * 2 + 2, tm + 1, tr, ind, value);
            }
            tree[x] = merge(tree[x * 2 + 1], tree[x * 2 + 2]);
        }
    }

public:
    SegmentTree(const vector<int>& elements) {
        n = elements.size();
        tree.resize(4 * n, {0, 0});
        build(elements, 0, 0, n - 1);
    }

    pair<int, int> getMin(int left, int right) {
        return min_cnt(0, 0, n - 1, left, right - 1);
    }

    void updateElement(int index, int value) {
        update(0, 0, n - 1, index, value);
    }
};

int main() {
    int n, m;
    scanf("%d %d", &n, &m);
    vector<int> lst(n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &lst[i]);
    }

    SegmentTree seg_tree(lst);

    for (int i = 0; i < m; i++) {
        int query;
        scanf("%d", &query);
        if (query == 1) {
            int index, value;
            scanf("%d %d", &index, &value);
            seg_tree.updateElement(index, value);
        } else if (query == 2) {
            int left, right;
            scanf("%d %d", &left, &right);
            auto result = seg_tree.getMin(left, right);
            printf("%d %d\n", result.first, result.second);
        }
    }

    return 0;
}

